### ISSUE TITLE
amp-img/amp-anim: provide warnings for overly large image dimensions

### ISSUE THREAD

**Issue by dvoytenko** [ampproject/amphtml#302]
E.g. 10000px-wide image is too large for most of devices.

**Comment by AaronCritchley** [comment]
I'll take this on, could I have some guidance on where this needs to be added? Hopefully I'll be able to take the next one without having to ask for guidance :smile:

**Comment by dvoytenko** [comment]
Unfortunately, the validator has not been open-sourced yet. Let me see if we can find a way to proceed with it.

**Comment by cramforce** [comment]
Does this really need to go into the validator?

The algorithm would basically be:

- Do nothing unless getMode().development is true
- Go through all amp-img on page. 
  - Get the effective src (what was picked after srcset is resolved)
  - Find a way to get its actual dimensions
  - Might not be possible. In that case: Create a new image with the src
  - Put it into `<body>` but in a way that cannot be seen without dimensions.
  - Wait for onload
  - Get actual dimensions of image
  - Compare those to the rendered dimensions
  - If they are too far (TBD) apart, call console.error with message

**Comment by Gregable** [comment]
@cramforce , I think the answer to if this should be in the validator or not depends on if we'd like pages with overly large dimensions to eventually not be allowed. If it's simply a permanent helpful suggestion, we probably don't want the validator to have it, as we expect that the validator will be used down the line in implementations of crawlers/renderers/etc. 

@powdercloud has expressed some good arguments for limiting permanent warnings in the validator as well due to the fact that they could cause a bit of blindness to temporary warnings and slow the ability to make changes down the line.

**Comment by cramforce** [comment]
@Gregable Right, I think this we can get this to not be crazy spammy but super helpful. I mostly want to tell people. I don't think anyone puts a 16 mega pixel image on a page on purpose, but if one made the mistake it is hard to notice if no one tells me.

I think the right place is a special "runtime suggestions" binary we load at the same time as the validator. @AaronCritchley you don't have to worry about this aspect :)

**Comment by dvoytenko** [comment]
Ok, sounds we are all clear then.

@AaronCritchley , you can start with this request. We will proceed as following:

1. We will initially host this code in https://github.com/ampproject/amphtml/blob/master/src/validator-integration.js. `maybeValidate` already checks for development mode.

2. We will initially assume the current device. We will extend it to a dictionary of typical devices in the future, but let's start simple. Device for us here is a width, height and DPR.

3. As @cramforce describes above, scan for `amp-img` (we may also do `amp-anim` later). For each `amp-img` that's currently not `display:none` (getBoundingClientRect().width > 0):

3.1. Resolve source URL based on the `src` and `srcset`. If desired you can expose a method in AmpImg class or just do it manually - either way. Notice you can only do it when you know the current width of the image (likewise will be improved later). Let's call it `layoutWidth`

3.2. Crate a new Image instance with src. Listen for load. Once loaded, look for the actual width/height of the image. I think it's `naturalWidth` or maybe just `width` should do. Let's call it `naturalWidth`

3.3. If `naturalWidth < layoutWidth * DPR / 2` - report warning "Image is too small"

3.4. If `naturalWidth > layoutWidth * DPR * 2` - report warning "Image is too large"

**Comment by AaronCritchley** [comment]
Ok, I'll get down to work, might take me longer than it should to do it right, hope you don't mind! 

If I have any questions or similar what's the best way to ask, via the Google group or do you guys plan to create an IRC for the project?

**Comment by dvoytenko** [comment]
Let's continue with the thread here. We're not quite ready to do IRC yet.

**Comment by rudygalfi** [comment]
Adding @ericlindley-g to feed into content quality project.

**Comment by cathyxz** [comment]
I think we don't need to do this anymore because lighthouse already has a check for image optimization and oversized images will result in docked points.
